src/middlewared/middlewared/plugins/smart.py

```diff
@@ -1,8 +1,148 @@
 import re
+from itertools import chain
 
-from middlewared.schema import accepts, Dict, Int, List, Str
-from middlewared.validators import Email
-from middlewared.service import SystemServiceService, private
+from middlewared.schema import accepts, Dict, Int, List, Patch, Str
+from middlewared.validators import Email, Range
+from middlewared.service import CRUDService, private, SystemServiceService, ValidationErrors
+
+
+class SMARTTestService(CRUDService):
+
+    class Config:
+        datastore = 'tasks.smarttest'
+        datastore_extend = 'smart.test.smart_test_extend'
+        datastore_prefix = 'smarttest_'
+        namespace = 'smart.test'
+
+    async def smart_test_extend(self, data):
+        disks = data.pop('disks')
+        data['disks'] = [disk['disk_identifier'] for disk in disks]
+        return data
+
+    @private
+    async def validate_data(self, data, schema):
+        verrors = ValidationErrors()
+
+        smart_tests = await self.query(filters=[('type', '=', data['type'])])
+        configured_disks = [d for test in smart_tests for d in test['disks']]
+        disks_dict = {disk['identifier']: disk['name'] for disk in (await self.middleware.call('disk.query'))}
+
+        disks = data.get('disks')
+        used_disks = []
+        invalid_disks = []
+        for disk in disks:
+            if disk in configured_disks:
+                used_disks.append(disks_dict[disk])
+            if disk not in disks_dict.keys():
+                invalid_disks.append(disk)
+
+        if used_disks:
+            verrors.add(
+                f'{schema}.disks',
+                f'The following disks already have tests for this type: {", ".join(used_disks)}'
+            )
+
+        if invalid_disks:
+            verrors.add(
+                f'{schema}.disks',
+                f'The following disks are invalid: {", ".join(invalid_disks)}'
+            )
+
+        return verrors
+
+    @accepts(
+        Dict(
+            'smart_task_create',
+            Str('desc'),
+            List('disks', items=[Str('disk')], required=True),
+            Str('dayweek'),
+            Str('month'),
+            Str('daymonth'),
+            Str('hour'),
+            Str('type', enum=['L', 'S', 'C', 'O'], required=True),
```

Comment:
This looks a bit cryptic, can we use more descriptive names, please? e.g. LONG, SHORT?